### PR TITLE
Customize PyTorch task error files

### PR DESCRIPTION
**Pull request by bgedik**
## Tracking issue
TODO

## Why are the changes needed?
TODO

## What changes were proposed in this pull request?
TODO
## How was this patch tested?
TODO

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 1]
```diff
@@ -334,6 +334,25 @@ def __init__(self, task_config: Elastic, task_function: Callable, **kwargs):
                 self.pod_template = PodTemplate()
             add_shared_mem_volume_to_pod_template(self.pod_template)
 
+    def get_error_file_name_suffix(self) -> Optional[str]:
+        """
+        Returns error file name suffix that represents the replica index.
+        """
+        # This is fragile, as it assumes hostname == podname, which is not the case if
```
Comment (bgedik) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:341]: What's needed is to customize the pod spec to set an environment variable that captures the value of metadata.labels."training.kubeflow.org/replica-index" :

```
env:
 - name: FLYTE_PYTORCH_TASK_REPLICA_INDEX
   valueFrom:
     fieldRef:
       fieldPath: metadata.labels['training.kubeflow.org/replica-index']
```

File: flytekit/core/base_task.py [hunk 2]
```diff
@@ -559,6 +559,15 @@ def task_config(self) -> Optional[T]:
         """
         return self._task_config
 
+    def get_error_file_name_suffix(self) -> Optional[str]:
```
Comment (fg91) [inline flytekit/core/base_task.py:562]: Nit (can be ignored right now until we are sure of how to implement this feature):
We shouldn't mention specific plugins like elastic here but rather speak of plugins using multiple workers which might produce independent error files.

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 3]
```diff
@@ -455,6 +474,14 @@ def fn_partial():
         except SignalException as e:
             logger.exception(f"Elastic launch agent process terminating: {e}")
             raise IgnoreOutputs()
+        except Exception as e:
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:477]: Can we do this in another PR please?
Is my understanding correct that any actual user error would have been a `ChildFailedError`? Do you have an example of an exception that isn't yet handled?

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 1 above)
Comment (fg91) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:341]: Setting this env var via the downward api in the pod spec in flyteplugins would in theory work I'd say 👍
I agree that we shouldn't use the hostname.

I have two questions though:

* Does the error file name actually need to contain the group rank? If the error message itself would contain the info which group rank failed, could we get away with a random uuid suffix which just prevents overwriting?
* Do you know how torch distributed elastic determines the group rank for the worker pods/processes? I just checked on a running elastic task and (in line with what you said) only see the `PET_NNODES` and `PET_NPROC_PER_NODE` env vars which specify the number of nodes and processes but not the rank. Could it be that `worker-0` ends up being group rank 2 or how is this decided at rendezvous?

Comment (bgedik) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:341]: The ranks are decided by the rendezvous and worker-0 is not necessarily the one that has rank 0 or group rank 0.

I agree that the suffix does not need to be the group rank. We can do uuid, but internal debugging might get a bit harder, as by looking at the error files, no one would know what pod actually generated it.

What about using the pod name? Is that available in Flyte via some mechanism or does that also require downward api usage?

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 3 above)
Comment (bgedik) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:477]: I've seen `RendezvousError` happening. Typically it happens when there is a much earlier ChildError from another group. Pod A has a ChildError, Pod B has a RendezvousError, and if Pod B's error file takes precedence, then we end up preventing a task restart :( 

I have this change in our local environment (via monkey patching `elastic_launch`), because with the current behavior of task outputs overwriting each other, these exceptions are preventing task restarts. 

I agree that this can belong to a separate PR. Btw, if we fix the task error file issue and pick the earliest, then the impact of this is smaller (but I still think treating these as non-recoverable user errors is a surprising behavior).

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 1 above)
Comment (fg91) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:341]: > The ranks are decided by the rendezvous and worker-0 is not necessarily the one that has rank 0 or group rank 0.

Ok then it makes sense 👍 

> What about using the pod name? Is that available in Flyte via some mechanism or does that also require downward api usage?

I think it's not available unfortunately. Could be added though I'd say.

> I agree that the suffix does not need to be the group rank. We can do uuid, but internal debugging might get a bit harder, as by looking at the error files, no one would know what pod actually generated it

The [`ChiledFailedError`](https://github.com/pytorch/pytorch/blob/f3df7deab8953af76ff1723ed49094208057a834/torch/distributed/elastic/multiprocessing/errors/__init__.py#L234) has a ref to the global rank but that doesn't help us because it's not linked to the pod and we are interested in the pod name in order to look at the respective stackdriver logs for more info - correct?

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 3 above)
Comment (fg91) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:477]: Can you please open another PR for this? I think we can merge this much faster than proper treatment of multiple error files.

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 1 above)
Comment (bgedik) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py]: `ChiledFailedError` is not the only error that comes out of `torch_elastic`, unfortunately. 

We can start with uuid and improve afterwards. Made the change.

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 3 above)
Comment (pingsutw) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py]: @fg91 Thanks for reviewing the PR 😄

**Comment by kumare3** [comment]
@bgedik did you give up on this?

**Comment by bgedik** [comment]
> @bgedik did you give up on this?

@kumare3 We can close this one. Currently working on backend changes on multi error file reader (there is an open PR). This PR will be reworked based on the latest RFC.

**Comment by bgedik** [comment]
> > @bgedik did you give up on this?
> 
> @kumare3 We can close this one. Currently working on backend changes on multi error file reader (there is an open PR). This PR will be reworked based on the latest RFC.

https://github.com/flyteorg/flytekit/pull/2797 is the new PR
